base_layer/core/src/base_node/tari_pulse_service/mod.rs

```diff
@@ -0,0 +1,190 @@
+        let timeout: Duration = Duration::from_secs(5);
+        let trust_anchor = Self::default_trust_anchor();
+
+        let (stream, handle) = TcpClientStream::<AsyncIoTokioAsStd<TokioTcpStream>>::new(([1, 1, 1, 1], 53).into());
+        let dns_muxer = DnsMultiplexer::<_, SigSigner>::with_timeout(stream, handle, timeout, None);
+        let (client, bg) = AsyncDnssecClient::builder(dns_muxer)
+            .trust_anchor(trust_anchor)
+            .build()
+            .await?;
+
+        tokio::spawn(bg);
+
+        Ok(client)
+    }
+
+    async fn run(
+        &mut self,
+        mut base_node_service: LocalNodeCommsInterface,
+        notify_passed_checkpoints: watch::Sender<bool>,
+    ) {
+        let mut interval = time::interval(self.config.check_interval.unwrap());
+        let mut interval_failed = time::interval(Duration::from_millis(100));
+        loop {
+            let dns_checkpoints = match self.fetch_checkpoints().await {
+                Ok(checkpoints) => checkpoints,
+                Err(e) => {
+                    error!(target: LOG_TARGET, "Error fetching DNS checkpoints: {:?}", e);
+                    interval_failed.tick().await;
+                    continue;
+                },
+            };
+
+            let heights = dns_checkpoints.iter().map(|(height, _)| *height).collect();
+            let local_checkpoints = match self.get_node_blocks(&mut base_node_service, heights).await {
+                Ok(checkpoints) => checkpoints,
+                Err(e) => {
+                    error!(target: LOG_TARGET, "Error fetching local checkpoints: {:?}", e);
+                    interval_failed.tick().await;
+                    continue;
+                },
+            };
+
+            let passed_checkpoints = dns_checkpoints.iter().all(|(height, hash)| {
+                local_checkpoints
+                    .iter()
+                    .any(|(local_height, local_hash)| height == local_height && hash == local_hash)
+            });
+            notify_passed_checkpoints.send(!passed_checkpoints).unwrap();
+            interval.tick().await;
+        }
+    }
+
+    async fn get_node_blocks(
+        &mut self,
+        base_node_service: &mut LocalNodeCommsInterface,
+        heights: Vec<u64>,
+    ) -> Result<Vec<(u64, String)>, anyhow::Error> {
+        let historical_blocks = future::try_join_all(heights.into_iter().map(|height| {
+            let mut node_clone = base_node_service.clone();
+            async move {
+                node_clone.get_header(height).await.and_then(|header| match header {
+                    Some(header) => Ok((header.height(), header.hash().to_hex())),
+                    None => {
+                        error!(target: LOG_TARGET, "Header not found for height: {}", height);
+                        Err(CommsInterfaceError::InternalError("Header not found".to_string()).into())
+                    },
+                })
+            }
+        }))
+        .await?;
```

Comment:
This is very weirdly written. 

But you only seem to care about a true or false here if you are on the chain or not. 
So all you have to do is check the last height. if the last height = true, then you are on the same chain.

Reply:
Rewritten logic so we only check for the newest block, reducing numbers of calls to get block header